x/yarpctest/helpers/port.go

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package helpers
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/yarpc/x/yarpctest/api"
+)
+
+// NewPortProvider creates an object that can be used to synchronize ports in
+// yarpctest infrastructure.  Ports can be acquired through the "Port" function
+// which will create new ports for the test based on the id passed into the
+// function.
+func NewPortProvider(t api.TestingT) *PortProvider {
+	return &PortProvider{
+		idToPort: make(map[string]*Port),
+		t:        t,
+	}
+}
+
+// PortProvider maintains a list of IDs to Ports.
+type PortProvider struct {
+	idToPort map[string]*Port
+	t        api.TestingT
+}
+
+// Port will return a *Port object that exists for the passed in 'id', or it
+// will create a *Port object if one does not already exist.
+func (p *PortProvider) Port(id string) *Port {
```

Comment:
Why a string?

Reply:
:man_shrugging: not that important.  String for the purpose of adding meaning to the port like "tchannel_port1" or something.

Reply:
Will that ever actually happen, like is there some concept of named ports that we have I haven't seen? Heh. I'd just do what it is, an int (or uint16), but alas.

Reply:
Heh note in your code example in the description, you used an int, which is probably because it felt natural to :-)

Reply:
So one of the confusions I want to avoid is thinking that the port provider is using an explicit port when it's not.  The PortProvider will always pick a random port by listening on `:0`.  But if I pass in a number like p.Port(12345) it seems like the port that will be listened on is `12345` (even though the actual port could be anything).  If we make the port a string we can decrease that confusion by having random text that could not be a port `p.Port("mytestname")`.

Also it's a bit confusing but there are actually two "Port" functions, the `PortProvider.Port` function and the top level `Port` function, they have different interfaces because the base `Port` function does not listen on `:0` to get a port, it uses the number it was given.  I'm open to suggestions to clear this up.

Reply:
To avoid the confusion, could name the method NamedPort. Port means it returns a *Port. Named gives meaning to the argument.

I’m also fine with it as-is.